scenarios/aws/microVMs/microvms/libvirt_fs.go

```diff
@@ -314,6 +314,31 @@ func downloadRootfs(fs *LibvirtFilesystem, runner *Runner, depends []pulumi.Reso
 	return waitFor, nil
 }
 
+func extractRootfs(fs *LibvirtFilesystem, runner *Runner, depends []pulumi.Resource) ([]pulumi.Resource, error) {
+	var waitFor []pulumi.Resource
+	for _, volume := range fs.volumes {
+		fsImage := volume.UnderlyingImage()
+
+		// Extract archive if it is xz compressed, which will be the case when downloading from remote S3 bucket.
+		// To do this we check if the magic bytes of the file at imagePath is fd377a585a00. If so then this is already
+		// a qcow2 file. No need to extract it. Otherwise it SHOULD be xz compressed file. Attempt to uncompress.
+		//
+		// Magic bytes of xz: fd377a585a00
+		// Magic bytes of qcow2: 514649fb
+		extractTopLevelArchive := command.Args{
+			Create: pulumi.Sprintf("if xxd -p -l 6 %[1]s | grep -E '^fd377a585a00$'; then mv %[1]s %[2]s && xz -d %[2]s; elif xxd -p -l 4 %[1]s | grep -E '^514649fb$'; then echo '%[1]s is qcow2 file'; else false; fi", fsImage.imagePath, fmt.Sprintf("%s.xz", fsImage.imagePath)),
```

Comment:
<!-- Suggestions propose improvements to the current subject. It’s important to be explicit and clear on what is being suggested and why it is an improvement. -->
**💬 suggestion**
You could move `fd377a585a00` and `514649fb` to a const to help readability. I would also store `fmt.Sprintf("%s.xz", fsImage.imagePath)` into a variable and then put it into the `Create` pulumi sprintf call